docs/References/Options.md
# Configuration Options

DeepDrill is configured by specifying one or more profiles as command line parameters. Each profile contains a certain number of key-value pairs organized in sections. A typical profile looks like this:

```INI
[image]
width = 960
height = 540

[perturbation]
tolerance = 1e-6
rounds = 50

[approximation]
coefficients = 5
```

Each option is briefly explained below: 

## Key-value pairs

### Section `[location]`

| <div style="width:120px">Key</div> | <div style="width:100px">Default value</div> | Description |
|-----|---------|-------------|
| `real`  | 0.0 | The real part of the image's center coordiate.
| `imag`  | 0.0 | The imaginary part of the image's center coordiate.
| `zoom`  | 1.0 | The magnification level of the image.
| `depth`  | 1.0 | The Maximum number of iterations.
| `escape`  | 1e32 | Once the iterated value gets bigger than this value, the corresponding coordinate is considered to be part of the Mandelbrot set.


### Section `[map]`

| <div style="width:120px">Key</div> | <div style="width:100px">Default value</div> | Description |
|-----|---------|-------------|
| `width`  | 1920 | The horizontal resolution of the drill map. If this key is not specified, the width of the drill map is automatically chosen to match the image width.
| `height`  | 1080 | The vertical resolution of the drill map. If this key is not specified, the width of the drill map is automatically chosen to match the image height.
| `depth`  | 1 | Possible values are 0 and 1. If set to 1, a normal map will be written to the `.map` file.
| `compress`  | yes | Indicates whether map files should be saved in compressed format.


### Section `[image]`

| <div style="width:120px">Key</div> | <div style="width:100px">Default value</div> | Description |
|-----|---------|-------------|
| `width`  | 1920 | The horizontal resolution of the image in pixels.
| `height`  | 1080 | The vertical resolution of the image in pixels.



### Section `[video]`

| <div style="width:120px">Key</div> | <div style="width:100px">Default value</div> | Description |
|-----|---------|-------------|
| `framerate`  | 60 | Frames per second of the computed video. By default, videos with 60 fps are computed to match the frame rate of standard TFTs.
| `keyframes`  | *derived* | Number of keyframes of the final video. If this key is not specified, DeepDrill derives the number from the zoom factor given in the location section.
| `startframe`  | 0 | Number of the keyframe where the video starts. 
| `velocity`  | 1.0 | Zoom speed. The assigned value defines the time interval between two keyframes in seconds.
| `bitrate`  | 8000 | This value is passed directly to the FFmpeg backend. Higher values result in better image quality.


### Section `[palette]`

| <div style="width:120px">Key</div> | <div style="width:100px">Default value</div> | Description |
|-----|---------|-------------|
| `image`  | - | Path to an image file storing the color palette. DeepDrill will derive the color palette from the uppermost horizontal line of the image.
| `scale`  | 1.0 | An optional scaling factor that can be used to stretch or shrink the color palette.  
| `offset`  | 0.0 | An optional offset that can be used to rotate the color palette.  


### Section `[texture]`
| <div style="width:120px">Key</div> | <div style="width:100px">Default value</div> | Description |
|-----|---------|-------------|
| `image`  | - | Path to an overlay image. If an overlay image is specified, the derived color value will be mixed with the pixels from the overlay image.
| `opacity`  | 0.5 | The weight factor used to mix color pixels and overlay pixels.
| `scale`  | 1.0 | An optional scaling factor that can be used to stretch or shrink the overlay image.
| `offset`  | 0.0 | An optional offset that can be used to rotate the overlay image.

### Section `[lighting]`
| <div style="width:120px">Key</div> | <div style="width:100px">Default value</div> | Description |
|-----|---------|-------------|
| `enable`  | no | Indicates whether the normal map should be utilized to create a spatial image.
| `alpha`  | 45 | The direction of the incoming light on the xy plane.
| `beta`  | 45 | The inclination of the light source.  


### Section `[gpu]`
| <div style="width:120px">Key</div> | <div style="width:100px">Default value</div> | Description |
|-----|---------|-------------|
| `colorizer`  | gradient.glsl | Path to a GLSL shader. The colorizer is the first of three GPU shaders. It utilzes the color palette to derive a RGB value for each pixel.
| `illuminator`  | lambert.glsl | Path to a GLSL shader. The illuminator is the second of three GPU shaders. It applies a lighting model to create a spatial images. This shader is skipped if lighting is disabled. 
| `scaler`  | tricubic.glsl | Path to a GLSL shader. The scaler is the third of three GPU shaders. It scales the internal image down to the final image.


### Section `[perturbation]`

| <div style="width:120px">Key</div> | <div style="width:100px">Default value</div> | Description |
|-----|---------|-------------|
| `enable`  | yes | Indicates if perturbation should be used to calculate the image. If set to false, DeepDrill falls back to the (very slow) standard algorithm. 
| `tolerance`  | 1e-6 | This value is used by the perturbation algorithm. Please refer to the *Theory* section for details. 
| `badpixels`  | 0.001 | Percentage of pixels that are allowed to be miscolored. 
| `rounds`  | 50 | This value is used by the perturbation algorithm. Please refer to the *Theory* section for details. 
| `color`  | black | Color used for colorizing glitch points.


### Section `[approximation]`

| <div style="width:130px">Key</div> | <div style="width:100px">Default value</div> | Description |
|-----|---------|-------------|
| `enable`  | yes | Indicates if series approximation should be used to calculate the image. If set to false, the iteration always starts at 0. Please refer to the *Theory* section for details.  
| `coefficients`  | 5 | This value is used by the series approximation algorithm. Please refer to the *Theory* section for details. 
| `tolerance`  | 1e-12 | This value is used by the series approximation algorithm. Please refer to the *Theory* section for details. 


### Section `[areachecking]`

| <div style="width:120px">Key</div> | <div style="width:100px">Default value</div> | Description |
|-----|---------|-------------|
| `enable`  | yes | Specifies whether each pixel is to be tested for belonging to the main bulb or the cartioid. If the test is positive, the pixel is known to belong to the Mandelbrot set without further calculations.
| `color`  | black | Color used for colorizing pixels with a positive area check.


### Section `[attractorcheck]`

| <div style="width:120px">Key</div> | <div style="width:100px">Default value</div> | Description |
|-----|---------|-------------|
| `enable`  | yes | Specifies whether attractor checking should be carried out. If the test is positive, the pixel is known to belong to the Mandelbrot set without further calculations.
| `tolerance`  | 1e-34 | Tolerance for considering two values as beeing equal. 
| `color`  | black | Color used for colorizing pixels with a positive attractor check.


### Section `[periodcheck]`

| <div style="width:120px">Key</div> | <div style="width:100px">Default value</div> | Description |
|-----|---------|-------------|
| `enable`  | yes | Specifies whether period checking should be carried out. If the test is positive, the pixel is known to belong to the Mandelbrot set without further calculations.
| `tolerance`  | 1e-74 | Tolerance for considering two values as beeing equal. 
| `color`  | black | Color used for colorizing pixels with a positive period check.


docs/_sources/References/Options.md.txt
# Configuration Options

DeepDrill is configured by specifying one or more profiles as command line parameters. Each profile contains a certain number of key-value pairs organized in sections. A typical profile looks like this:

```INI
[image]
width = 960
height = 540

[perturbation]
tolerance = 1e-6
rounds = 50

[approximation]
coefficients = 5
```

Each option is briefly explained below: 

## Key-value pairs

### Section `[location]`

- `real`

  Real part of the image's center coordiate.

- `imag`

  Imaginary part of the image's center coordiate.

- `zoom`

  The magnification level of the image.

- `depth`

  Maximum number of iterations.

### Section `[map]`

- `width`
  
  The horizontal resolution of the drill map. If this key is not specified, the width of the drill map is automatically chosen to match the image width.

- `height` 

  The vertical resolution of the drill map. If this key is not specified, the width of the drill map is automatically chosen to match the image height.

- `depth` 

  Possible values are 0 and 1. If set to 1, a normal map will be written to the `.map` file.

### Section `[image]`

- `width`
  
  The horizontal resolution of the image.

- `height` 

  The vertical resolution of the image.

- `depth` 

  Possible values are 0 and 1. If set to 1, a spatial effect will be generated from the normal map. 

- `illuminator` 

  Path to a GLSL shader. The illuminator is the first of two GPU shaders. It derives the RGB value for each pixel and applies a lighting model.

- `scaler` 

  Path to a GLSL shader. The scaler is the second of two GPU shaders. It scales the internal image down to the final image.

### Section `[video]`

- `framerate`

  Frames per second of the computed video. By default, videos with 60 fps are computed to match the frame rate of standard TFTs.

- `keyframes`

  Number of keyframes of the final video. If this key is not specified, DeepDrill derives the number from the zoom factor given in the location section.

- `inbetweens`

  Number of interpolated frames between two keyframes. By default, this value is assigned twice the frame rate. Consequently, each keyframe is displayed for 2 seconds.

- `bitrate`
  
  This value is passed directly to the FFmpeg backend. Higher values result in better image quality.

- `scaler` 

  Path to a GLSL shader. If a video is created, this downscaler is used instead of the one specified in the image section. 

### Section `[colors]`

- `palette`

  Path to an image file storing the color palette. DeepDrill will derive the color palette from the uppermost horizontal line of the image. 

- `scale`

  An optional scaling factor that can be used to stretch or shrink the color palette.  

- `alpha`, `beta`

  These two values define the direction of the light vector. They only take effect if spatial images are computed, i.e., if `depth` is set to 1 in the image section. 

### Section `[perturbation]`

- `enable`

  Indicates if perturbation should be used to calculate the image. If set to false, DeepDrill falls back to the (very slow) standard algorithm. 

- `tolerance`

  This value is used by the perturbation algorithm. Please refer to the *Theory* section for details. 

- `badpixels`

  Percentage of pixels that are allowed to be miscolored.

- `rounds`

  This value is used by the perturbation algorithm. Please refer to the *Theory* section for details. 

### Section `[approximation]`

- `enable`

  Indicates if series approximation should be used to calculate the image. If set to false, the iteration always starts at 0. Please refer to the *Theory* section for details.  

- `coefficients`

  This value is used by the series approximation algorithm. Please refer to the *Theory* section for details. 

- `tolerance`

  This value is used by the series approximation algorithm. Please refer to the *Theory* section for details. 

### Section `[debug]`

- `glitches`

  If set to `true`, glitch points are marked red in the final image. It's main purpose is to debug the glitch detection algorithm.
  

src/shared/Options.cpp
// -----------------------------------------------------------------------------
// This file is part of DeepDrill
//
// A Mandelbrot generator based on perturbation and series approximation
//
// Copyright (C) Dirk W. Hoffmann. www.dirkwhoffmann.de
// Licensed under the GNU General Public License v3
//
// See https://www.gnu.org for license information
// -----------------------------------------------------------------------------

#include "Options.h"
#include "AssetManager.h"
#include "Coord.h"
#include "IO.h"
#include "Logger.h"
#include "Parser.h"

namespace dd {

bool Options::stop;

Options::Flags Options::flags;
Options::Files Options::files;
Options::Location Options::location;
Options::Drillmap Options::drillmap;
Options::Mapfile Options::mapfile;
Options::Image Options::image;
Options::Video Options::video;
Options::Palette Options::palette;
Options::Distance Options::distance;
Options::Texture Options::texture;
Options::Lighting Options::lighting;
Options::GPU Options::gpu;
Options::Perturbation Options::perturbation;
Options::Approximation Options::approximation;
Options::Areacheck Options::areacheck;
Options::Attractorcheck Options::attractorcheck;
Options::Periodcheck Options::periodcheck;

std::map<string,string> Options::keys;
std::vector<string> Options::overrides;
std::map<string,string> Options::defaults = [](){

    std::map<string,string> defaults;

    // Location keys
    defaults["location.real"] = "0.0";
    defaults["location.imag"] = "0.0";
    defaults["location.zoom"] = "1.0";
    defaults["location.depth"] = "800";
    defaults["location.escape"] = "1e32";

    // Map keys
    defaults["map.width"] = "1920";
    defaults["map.height"] = "1080";

    // Mapfile keys
    defaults["mapfile.compress"] = "yes";
    defaults["mapfile.result"] = "yes";
    defaults["mapfile.first"] = "yes";
    defaults["mapfile.last"] = "yes";
    defaults["mapfile.nitcnt"] = "yes";
    defaults["mapfile.derivative"] = "no";
    defaults["mapfile.normal"] = "yes";
    defaults["mapfile.dist"] = "yes";

    // Image keys
    defaults["image.width"] = "1920";
    defaults["image.height"] = "1080";

    // Video keys
    defaults["video.framerate"] = "60";
    defaults["video.keyframes"] = "0";
    defaults["video.startframe"] = "0";
    defaults["video.velocity"] = "1.0";
    defaults["video.bitrate"] = "8000";

    // Palette keys
    defaults["palette.image"] = "";
    defaults["palette.bgcolor"] = "black";
    defaults["palette.mode"] = "smooth";
    defaults["palette.scale"] = "1.0";
    defaults["palette.offset"] = "-0.15";

    // Distance estimation keys
    // defaults["distance.enable"] = "no";
    defaults["distance.threshold"] = "0.2";

    // Texture keys
    defaults["texture.image"] = "";
    defaults["texture.opacity"] = "0.5";
    defaults["texture.scale"] = "1.0";
    defaults["texture.offset"] = "0.0";

    // Lighting keys
    defaults["lighting.enable"] = "no";
    defaults["lighting.alpha"] = "45";
    defaults["lighting.beta"] = "45";

    // GPU keys
    defaults["gpu.colorizer"] = "gradient.glsl";
    defaults["gpu.illuminator"] = "lambert.glsl";
    defaults["gpu.scaler"] = "tricubic.glsl";

    // Perturbation keys
    defaults["perturbation.enable"] = "yes";
    defaults["perturbation.tolerance"] = "1e-6";
    defaults["perturbation.badpixels"] = "0.001";
    defaults["perturbation.rounds"] = "50";
    defaults["perturbation.color"] = "";

    // Approximation keys
    defaults["approximation.enable"] = "yes";
    defaults["approximation.coefficients"] = "5";
    defaults["approximation.tolerance"] = "1e-12";

    // Area checking keys
    defaults["areacheck.enable"] = "yes";
    defaults["areacheck.color"] = "";

    // Attractor checking keys
    defaults["attractorcheck.enable"] = "yes";
    defaults["attractorcheck.tolerance"] = "1e-34";
    defaults["attractorcheck.color"] = "";

    // Period checking keys
    defaults["periodcheck.enable"] = "yes";
    defaults["periodcheck.tolerance"] = "1e-74";
    defaults["periodcheck.color"] = "";

    return defaults;
}();

std::vector <fs::path>
Options::getInputs(Format format)
{
    std::vector <fs::path> result;

    for (const auto &it : files.inputs) {
        if (AssetManager::getFormat(it) == format) result.push_back(it);
    }
    return result;
}

std::vector <fs::path>
Options::getOutputs(Format format)
{
    std::vector <fs::path> result;

    for (const auto &it : files.outputs) {
        if (AssetManager::getFormat(it) == format) result.push_back(it);
    }
    return result;
}

void
Options::parse(string keyvalue)
{
    // Seek the '=' symbol
    auto pos = keyvalue.find("=");
    if (pos == std::string::npos) throw SyntaxError("Parse error");

    // Split string
    auto key = keyvalue.substr(0, pos);
    auto value = keyvalue.substr(pos + 1, std::string::npos);

    parse(key, value);
}

void
Options::parse(string key, string value)
{
    // Convert the key to lower case
    std::transform(key.begin(), key.end(), key.begin(),
                   [](unsigned char ch){ return std::tolower(ch); });

    // Remove white spaces
    erase(value, ' ');

    // Strip quotation marks
    value.erase(remove(value.begin(), value.end(), '\"'), value.end());

    keys[key] = value;

    try {

        if (key == "location.real") {

            Parser::parse(value, location.real);

        } else if (key == "location.imag") {

            Parser::parse(value, location.imag);

        } else if (key == "location.zoom") {

            Parser::parse(value, location.zoom);

        } else if (key == "location.depth") {

            Parser::parse(value, location.depth);

        } else if (key == "location.escape") {

            Parser::parse(value, location.escape);

        } else if (key == "map.width") {

            Parser::parse(value, drillmap.width, MIN_MAP_WIDTH, MAX_MAP_WIDTH);

        } else if (key == "map.height") {

            Parser::parse(value, drillmap.height, MIN_MAP_HEIGHT, MAX_MAP_HEIGHT);

        } else if (key == "mapfile.compress") {

            Parser::parse(value, mapfile.compress);

        } else if (key == "mapfile.result") {

            Parser::parse(value, mapfile.result);

        } else if (key == "mapfile.first") {

            Parser::parse(value, mapfile.first);

        } else if (key == "mapfile.last") {

            Parser::parse(value, mapfile.last);

        } else if (key == "mapfile.nitcnt") {

            Parser::parse(value, mapfile.nitcnt);

        } else if (key == "mapfile.derivative") {

            Parser::parse(value, mapfile.derivative);

        } else if (key == "mapfile.normal") {

            Parser::parse(value, mapfile.normal);

        } else if (key == "mapfile.dist") {

            Parser::parse(value, mapfile.dist);

        } else if (key == "image.width") {

            Parser::parse(value, image.width, MIN_IMAGE_WIDTH, MAX_IMAGE_WIDTH);

        } else if (key == "image.height") {

            Parser::parse(value, image.height, MIN_IMAGE_HEIGHT, MAX_IMAGE_HEIGHT);

            if (image.height % 2 == 1) {
                throw Exception("Height must be dividable by 2");
            }

        } else if (key == "video.framerate") {

            Parser::parse(value, video.frameRate, 25, 240);

        } else if (key == "video.keyframes") {

            Parser::parse(value, video.keyframes);

        } else if (key == "video.startframe") {

            Parser::parse(value, video.startframe);

        } else if (key == "video.velocity") {

            Parser::parse(value, video.velocity);

        } else if (key == "video.bitrate") {

            Parser::parse(value, video.bitrate);

        } else if (key == "palette.image") {

            Parser::parse(value, palette.image, { Format::BMP, Format::JPG, Format:: PNG });

        } else if (key == "palette.bgcolor") {

            Parser::parse(value, palette.bgColor);

        } else if (key == "palette.mode") {

            Parser::parse(value, palette.mode);

        } else if (key == "palette.scale") {

            Parser::parse(value, palette.scale);

        } else if (key == "palette.offset") {

            Parser::parse(value, palette.offset);

        /*
        } else if (key == "distance.enable") {

            Parser::parse(value, distance.enable);
        */

        } else if (key == "distance.threshold") {

            Parser::parse(value, distance.threshold);

        } else if (key == "texture.image") {

            Parser::parse(value, texture.image, { Format::BMP, Format::JPG, Format:: PNG });

        } else if (key == "texture.opacity") {

            Parser::parse(value, texture.opacity);

        } else if (key == "texture.scale") {

            Parser::parse(value, texture.scale);

        } else if (key == "texture.offset") {

            Parser::parse(value, texture.offset);

        } else if (key == "lighting.enable") {

            Parser::parse(value, lighting.enable);

        } else if (key == "lighting.alpha") {

            Parser::parse(value, lighting.alpha);

        } else if (key == "lighting.beta") {

            Parser::parse(value, lighting.beta);

        } else if (key == "gpu.colorizer") {

            Parser::parse(value, gpu.colorizer, { Format::GLSL });

        } else if (key == "gpu.illuminator") {

            Parser::parse(value, gpu.illuminator, { Format::GLSL });

        } else if (key == "gpu.scaler") {

            Parser::parse(value, gpu.scaler, { Format::GLSL });

        } else if (key == "areacheck.enable") {

            Parser::parse(value, areacheck.enable);

        } else if (key == "areacheck.color") {

            Parser::parse(value, areacheck.color);

        } else if (key == "attractorcheck.enable") {

            Parser::parse(value, attractorcheck.enable);

        } else if (key == "attractorcheck.tolerance") {

            Parser::parse(value, attractorcheck.tolerance);

        } else if (key == "attractorcheck.color") {

            Parser::parse(value, attractorcheck.color);

        } else if (key == "periodcheck.enable") {

            Parser::parse(value, periodcheck.enable);

        } else if (key == "periodcheck.tolerance") {

            Parser::parse(value, periodcheck.tolerance);

        } else if (key == "periodcheck.color") {

            Parser::parse(value, periodcheck.color);

        } else if (key == "perturbation.enable") {

            Parser::parse(value, perturbation.enable);

        } else if (key == "perturbation.tolerance") {

            Parser::parse(value, perturbation.tolerance);

        } else if (key == "perturbation.badpixels") {

            Parser::parse(value, perturbation.badpixels);

        } else if (key == "perturbation.rounds") {

            Parser::parse(value, perturbation.rounds);

        } else if (key == "perturbation.color") {

            Parser::parse(value, perturbation.color);

        } else if (key == "approximation.enable") {

            Parser::parse(value, approximation.enable);

        } else if (key == "approximation.coefficients") {

            Parser::parse(value, approximation.coefficients);

        } else if (key == "approximation.tolerance") {

            Parser::parse(value, approximation.tolerance);

        } else {

            throw Exception("Unknown key");
        }

    } catch (Exception &e) {
        throw KeyValueError(key, e.what());
    }
}

void
Options::applyDefaults()
{
    // Adjust some default values
    if (keys.find("image.width") != keys.end()) {
        defaults["map.width"] = keys["image.width"];
    }
    if (keys.find("image.height") != keys.end()) {
        defaults["map.height"] = keys["image.height"];
    }

    // Use default values for all missing keys
    for (auto &it : defaults) {
        if (keys.find(it.first) == keys.end()) {
            parse(it.first, it.second);
        }
    }

    // Apply overrides
    for (auto &it : overrides) {
        parse(it);
    }
}

void
Options::derive()
{
    // Derive unspecified video parameters
    if (!video.keyframes) {

        auto zoom = ExtendedDouble(location.zoom);
        video.keyframes = isize(std::ceil(zoom.log2().asDouble()));
    }
}

}
